index.js

```diff
@@ -248,13 +185,88 @@ function got(url, opts, cb) {
 		}
 
 		req.end();
-	}
+	});
+
+	ee.on('response', function (res) {
+		proxy.setReadable(res);
+		proxy.emit('response', res);
+	});
 
-	get(opts, cb);
+	ee.on('redirect', proxy.emit.bind(proxy, 'redirect'));
 
 	return proxy;
 }
 
+function normalizeArguments(url, opts) {
+	if (typeof url !== 'string' && typeof url !== 'object') {
+		throw new GotError('Parameter `url` must be a string or object, not ' + typeof url);
+	}
+
+	opts = objectAssign(
+		{
+			protocol: 'http:'
+		},
+		typeof url === 'string' ? urlLib.parse(prependHttp(url)) : url,
+		opts
+	);
+
+	opts.headers = objectAssign({
+		'user-agent': 'https://github.com/sindresorhus/got',
+		'accept-encoding': 'gzip,deflate'
+	}, lowercaseKeys(opts.headers));
+
+	if (opts.pathname) {
+		opts.path = opts.pathname;
+	}
+
+	var query = opts.query;
+	if (query) {
+		if (typeof query !== 'string') {
+			opts.query = querystring.stringify(query);
+		}
+
+		opts.path = opts.pathname + '?' + opts.query;
+		delete opts.query;
+	}
+
+	if (opts.json) {
+		opts.headers.accept = opts.headers.accept || 'application/json';
+	}
+
+	var body = opts.body;
+	if (body) {
+		if (typeof body !== 'string' && !Buffer.isBuffer(body) && !isStream.readable(body)) {
+			throw new GotError('options.body must be a ReadableStream, string or Buffer');
+		}
+
+		opts.method = opts.method || 'POST';
+
+		if (!opts.headers['content-length'] && !opts.headers['transfer-encoding'] && !isStream.readable(body)) {
+			var length = typeof body === 'string' ? Buffer.byteLength(body) : body.length;
+			opts.headers['content-length'] = length;
+		}
+	}
+
+	opts.method = opts.method || 'GET';
+
+	return opts;
+}
+
+function got(url, opts, cb) {
+	if (typeof opts === 'function') {
+		cb = opts;
+		opts = {};
+	}
+
+	opts = normalizeArguments(url, opts);
+
+	if (cb) {
+		return asCallback(opts, cb);
```

Comment:
I thought it was nice to read - return as callback, but to keep things consistent :+1: